### ISSUE TITLE
Allow miners to set expiration duration for sectors

### ISSUE THREAD

**Issue by jennijuju** [filecoin-project/lotus#4760]
The sectors expiration duration can be any time from 180 days - 540 days. Currently, all the cc sectors' expiration duration is set to 540 days by default, and we do not expose any cli for miners to customize it. Sectors with deals are expired when the last deal is ended.

Ideally, we should allow miners to set the below values in the configuration:

- [x] #6836

- [ ] minimum expiration duration for deal sectors before sealing

**Comment by momack2** [comment]
Seems like the two potential ways to design this are to either allow miners to explicitly pass a duration to override the default when sealing a new sector - or expose the default in something like the config for miners to configure within the accepted min/max bounds. 

Note, choosing a **shorter** duration than the default/max means that miners will have to reseal the sector to maintain power (expending work that could have gone toward sealing a new sector) - so the long-term expectation for rewards for unit effort is worse the shorter you configure your expiration duration.

**Comment by magik6k** [comment]
Roughly what needs to be done
* Add new fields to the Sealing config - `node/config`
* Plumb sealer config to https://github.com/filecoin-project/lotus/blob/master/extern/storage-sealing/precommit_policy.go
* Handle those in BasicPreCommitPolicy

**Comment by placer14** [comment]
Naming things is hard. Welcoming preferences on names:
```
type SealingConfig struct {
...
	// MinDealDurationInSector is the minimum required duration for any deal within a sector
	// before it can be sealed. This value is useful to control the minimum desired sector
	// lifetime as sectors expire when the longest deal is completed/terminated. (Note: Deals
	// which terminate prematurely could also cause the sector to expire prematurely if it is
	// the oldest deal remaining in the sector.) Value must be between 180-540 days inclusive.
	MinDealDurationInSector Duration

	// CommittedCapacityDefaultLifetime is the default duration a Committed Capacity (CC)
	// sector will live before it is either expired or converted into sector containing deals.
	// Value must be between 180-540 days inclusive.
	CommittedCapacityDefaultLifetime Duration
...
}
```

/cc @magik6k @jennijuju

**Comment by placer14** [comment]
@magik6k There's new implicit behavior in this scope: Currently the behavior only returns errors when the system fails to produce a value due to API invariants and does not seem to consider the new scenario where all pieces passed into `Expiration` have EndEpoch which is less than `MinDealDurationInSector`. Assumedly the worker should delay the sealing of this sector until more deals are batched into it but it's not clear how to signal this to the worker.

Should Expiration return an error in this case? Or should it only consider the pieces passed in and let the caller protect for the above case with an error returned within `Sealing.handlePreCommitting` in `extern/storage-sealing/states_sealing.go`?

**Comment by jennijuju** [comment]
MinDealDurationInSector is a bit confusing imho. AFAIK Sector expiration is not bounded by the longest deal duration, but the latest end epoch of the deals in the sector.

**Comment by jennijuju** [comment]
For CommittedCapacityDefaultLifetime, I will add `or extended lifetime` to *before it is either expired or converted into sector containing deals.*?

**Comment by placer14** [comment]
@jennijuju 
> ...but the latest end epoch of the deals in the sector.
Yes, correct. This is what I'm trying to capture in the name. This value intends to represent the "minimum deal duration that we would expect in the sector before it is sealed" and not the "longest deal duration possible by the protocol".
> I will add or extended lifetime to before it is either expired or converted into sector containing deals.?

What about this:
```
	// CommittedCapacityDefaultLifetime is the default duration a Committed Capacity (CC)
	// sector will live before it must be extended or converted into sector containing deals
        // before it is terminated.
	// Value must be between 180-540 days inclusive.
	CommittedCapacityDefaultLifetime Duration
```

**Comment by Kubuxu** [comment]
Names seem alright to me.


I'm not sure what to do about `handlePreCommitting` miners might want to refuse to accept deals that don't satisfy `MinDealDurationInSector`.
`handlePreCommiting` is too late for deciding or delaying deals, that decision needs to be made before sealing the sector.

**Comment by placer14** [comment]
I'm still not sure what to do to proceed with `MinDealDurationInSector`. The crux of my block is handling these params within BasicPreCommitPolicy. Am I intended to error out of BasicPreCommitPolicy.Expiration? Does handlePreCommitting handle that? (Nothing that Kuba believes it's too late to delay sealing from within handlePreCommitting. (And are there any side-effects to cleanup? If so, how?)

Note that @magik6k has also asked for validation of these values to be handled on lazy-load within the Policy, which is only instantiated from within `handlePreCommitting` so there is no way to validate the value before it fails to be used at least once (going to be poor UX when misconfigured).

Looking for some clarity on handling these rough edges before proceeding (and I believe this is context that only @magik6k can provide at the moment?).

**Comment by magik6k** [comment]
The way I see this:
* PreCommitPolicy is executed to get a sector expiration for a sector that is about to be precommitted, but was already sealed (and it may contain deals), so it has some value to the miner
* So it should try to handle most error cases as gracefully as possible - e.g. if user-configured values are out of bounds, it should use whatever value makes most sense withing the bounds, and log a warning. 
* If user configuration is totally invalid, also log a warning, and fallback on defaults (which should give the user most flexibility, so in this case the default should be the shortest possible sector lifetime, so in case of misconfiguration, those sectors can be extended after the fact (if the default would be higher lifetime, you can't shorten it after the sector is committed))

**Comment by Kubuxu** [comment]
@magik6k parameter being introduced here affects the scheduling of *sealing* of the sector with deals. So it cannot be handled in PreCommitPolicy.

**Comment by placer14** [comment]
(Note: I pulled the CC configuration into it's own [issue](https://github.com/filecoin-project/lotus/issues/6836)/[PR](https://github.com/filecoin-project/lotus/pull/6803). Leaving this issue to continue discussion of the proposed `MinDealDurationInSector` param.)

**Comment by placer14** [comment]
The scenario we're trying to handle more gracefully is the following: A worker has `MinDealDurationInSector` set to, say, `360 * 24 * time.Hour` (360 days) and the miner begins sealing a sector where none of the included deal pieces have an `EndEpoch` which extends out to `currentHeight + durationToEpochs(MinDealDurationInSector)`. I don't think we're protecting for this case _before_ `handlePreCommitting` and handling it _within_ `handlePreCommitting` (the first place `BasicPreCommitPolicy` (BPCP) is instantiated) will leave BPCP with the only resort of returning an error and failing the sealing sector in-progress. It sounds like we don't want this and will need to identify where (earlier in the sealing pipeline) we should be validating that a given sector satisfies the BPCP before starting to seal anything.

/cc @jennijuju

**Comment by placer14** [comment]
I believe this was closed prematurely. The PR only allows configurable CC sector expiration. This issue still has configurable non-CC sector expiration as outstanding scope.

**Comment by Meatball13** [comment]
I'd like to throw in the suggestion that it'd be nice to allow miners to set a minimum/maximum seal duration they'll accept for deals as well.  Right now with lotus-miner storage-deals set-ask you can set prices and min/max piece size, but you can't set Expiry.  Maybe some miners only want to hold short term deals, while others want longer term deals.
